tests/device_compilation_tests.cpp

```diff
@@ -25,82 +25,98 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-#include <initializer_list>
-
-#include <CL/sycl.hpp>
+#include "unit_tests.hpp"
 
-// What we're testing here is different forms of function calls, including
-// implicit and templated functions and constructors as well as destructors.
-// The hipSYCL Clang plugin must recognize that each of these is being called
-// by the user-provided kernel functor, and mark them as __device__ functions
-// accordingly.
-//
-// Additionally, we test that complex kernel names are correctly supported.
-
-void foo4(int v) {
-  printf("foo4: %d\n", v);
-}
-
-struct Conversion {
-  Conversion(float value) : value(value) {
-    foo4(*this);
-  }
-
-  operator int() {
-    return static_cast<int>(value);
-  }
+#include <initializer_list>
+#include <functional>
 
-  float value;
-};
+// NOTE: Since these tests are concerned with device compilation, the first
+// and arguably most important "test" is that this file compiles at all. We
+// then also include a few unit tests to assert the correct behavior of the
+// resulting programs.
 
-void foo3(int v) {
-  printf("foo3: %d\n", v);
-}
+BOOST_FIXTURE_TEST_SUITE(device_compilation_test_suite, reset_device_fixture)
 
-void foo2(Conversion foo) {
-  printf("foo2: %f\n", foo.value);
+int add_five(int v) {
+  return v + 5;
 }
 
-void foo1(int v) {
-  printf("foo1: %d\n", v + 5);
-  ([=](){ foo3(v); })();
+void call_lambda(int& v) {
+  ([&](){ v = add_five(v); })();
 }
 
 struct MyParent {
   MyParent(std::initializer_list<int> init) {
     for(auto&& e : init) {
       parent_value += e;
     }
-    foo1(parent_value);
+    call_lambda(parent_value);
   }
 
   virtual ~MyParent() {
-    printf("~MyParent: %d\n", parent_value);
+    call_on_dtor(parent_value + 7);
```

Comment:
When running auto-test framework (e.g., `HIPSYCL_NO_RUNTIME_TESTING=rocm python3 build.py psalz improve-shmem-detection`), device compilation tests fail on ROCm:

```
error: <unknown>:0:0: in function _ZN29device_compilation_test_suite8MyParentD2Ev void (%"struct.device_compilation_test_suite::MyParent"*): unsupported indirect call to function <unknown>
```
I'm not sure if this is a new problem introduced by your changes. Could well be that the more elaborate tests now expose an already existing problem. I'm also not sure if an `std::function` in kernel code is allowed by the spec. In principle, the spec doesn't really specify the interaction between STL and SYCL. I could well imagine that `std::function` breaks some assumptions (that are legal to make for SYCL) of device compilers that compile for architectures without stack/function pointers/proper functions and have to inline everything.
In general, I think we should make sure that the tests always run, and potentially change the tests.

Reply:
Ah yes, I didn't consider that the way `std::function` works might be illegal on some architectures. Luckily it's not integral to the test, I just wanted to assert somehow that the destructor of `MyParent` was correctly being called. I changed the mechanism - please try again!